Marlin/src/HAL/STM32/msc_sd.cpp

```diff
@@ -58,39 +64,65 @@ class Sd2CardUSBMscHandler : public USBMscHandler {
     // single block
     if (blkLen == 1) {
       hal.watchdog_refresh();
-      sd2card->writeBlock(blkAddr, pBuf);
-      return true;
+      return sd2card->writeBlock(blkAddr, pBuf);
     }
 
     // multi block optimization
-    sd2card->writeStart(blkAddr, blkLen);
-    while (blkLen--) {
+    uint32_t multi_retry_cnt = SD_MULTIBLOCK_RETRY_CNT;
+
+  RETRY_MULTI:
+    uint32 i = blkLen;
+    uint8_t *cBuf = pBuf;
+    sd2card->writeStart(blkAddr);
+    while (i--) {
       hal.watchdog_refresh();
-      sd2card->writeData(pBuf);
-      pBuf += BLOCK_SIZE;
+      if (sd2card->writeData(cBuf) == false)
+      {
+        sd2card->writeStop();
+        if (--multi_retry_cnt == 0)
+          goto FAIL;
+        multi_retry_cnt--;
+        goto RETRY_MULTI:
```

Comment:
NOTE − Use of goto statement is highly discouraged in any programming language because it makes difficult to trace the control flow of a program, making the program hard to understand and hard to modify. Any program that uses a goto can be rewritten to avoid them.

Reply:
Thanks for pointing that out. I had the same argument with @thinkyhead over this piece of code. But I think it is fine because the meaning of the code has been properly explained in the video. Also if goto were so bad then it would be deprecated in the language. Could you please look over this, ignore that people lectured you over evil goto? It is fine imo.

Reply:
I'm always open to different coding styles, but I'm also happy to mess around and see if I can make code that includes a 'goto' more "elegant" — and possibly fall on my face in the attempt. So, I went ahead and made some attempts, and I did "improve" the code in the sense that it results in a more top-down flow. In particular, I did re-add some `bool` flags and wrapped some code in `do { ... } while(0)` so that `break` can replace `goto`.

The results are not bad, so then the only question is whether it requires more or less cycles, and whether this is a bottleneck area of the code that those extra cycles matter. Marlin does spend a lot of time in the SD code path, but not to the extent that it becomes a significant bottleneck. As long as it can read as fast or faster than the typical serial port, I think we are safe.

Reply:
I am also open to equivalent code transformation! As long as the code performs 100% the same and removes doubt from co-devs.